Update various sending commands to return a value

Wolfsblvt opened the pull request:
### Changes
- Most send commands return the message being send now
  - `/sendas`
  - `/send`
  - `/sys`
  - `/comment`
  - `/ask` did already work this way
  - Can't be changed: `/trigger`, because by default it does not await the message generation
- Refactored checks on `/sendas` at the beginning, as they were very redundant
- Updated `/ask` toasts to warnings, being more consistent with the other commands.
  Errors should be reserved for something unexpectedly being wrong or actually failing. Missing or falsely formatted args can just be a warning.
- Function `sendMessageAsUser` now has a return value of the message object itself. Does not break existing code.

#### Possible "Breaking" Changes
`/sendas name={{char}} Some message beginning | /continue`
will not work anymore. It does not error out either, it'll silently pass the message as a custom prompt into the continue.
Solution is easy though, people should get used to using pipe breakers.
This works: `/sendas name={{char}} Some message beginning || /continue`

## Checklist:

- [x] I have read the [Contributing guidelines](https://www.youtube.com/watch?v=dQw4w9WgXcQ).

Wolfsblvt committed:
Refactor if checks on /sendas

Wolfsblvt committed:
Update return values of send commands

- /sendas
- /sys
- /comment

Wolfsblvt committed:
/send with return value too

- /send
- return message on `sendMessageAsUser` now

Wolfsblvt committed:
Update /ask toasts to warnings for consistency

Succubyss commented:
I'm personally in favor. However, if this is refused, let the ability to have it return a value at least become a parameter like `emit=` or `pass=`

Cohee1207 reviewed (changes requested):
This is a breaking change and I'm not a fan of it. This has limited usability (only ever usable if you have specific message-altering regex), but high muscle memory breakage potential. I'm ready to take a Luccy pill and suggest to enable this behavior with a named argument.

Wolfsblvt commented:
What would that one be named?
And should be update existing commands that are already returning to so this as well?

I think `return=` is the most reasonable actually, that we can consistently use across all commands in the future, without conflicts or being weirdly named for the command.
`output=`, like the current `listinjects` might be an option too, but I have a feeling that might go against some commands where output does not make much sense.

Values could be:
- none
- pipe
- toast
- console

Dunno if the options "chat" and "popup" would globally make sense, that one applies mostly to injects.

Cohee1207 commented:
> What would that one be named?

"return" is fine.

> And should be update existing commands that are already returning to so this as well?

Yes, but it's only /ask so far? + potential aliases and refactoring for /listinjects and /listvar

> output=, like the current listinjects might be an option too, but I have a feeling that might go against some commands where output does not make much sense.

Actually it's "format", not "output".  If you think this name sucks, let's add a couple aliases. And extract common function for doing the "return" thing.

Wolfsblvt commented:
> Actually it's "format", not "output". If you think this name sucks, let's add a couple aliases. And extract common function for doing the "return" thing.

Working on my fix/update. Gimme a bit. Though just wanted to reply, aliases for named aren't really well supported in the current implementation really. You'd have to build them yourself in the callback handler, which can be quite confusing. But it certainly is an option. Len said he does not want to have those though.

Wolfsblvt committed:
Update return types as optional via named arg

- Update the modified slash commands for chat sending to use the named arg
- Add `slashCommandReturnHelper` for shared funcitonality on return type usage

Wolfsblvt committed:
Update /ask with return types, defaulting 'pipe'

Wolfsblvt committed:
Refactor /listinjects, deprecate its "format" arg

Wolfsblvt committed:
Refactor /listvar, deprecate its "format" arg

- Update /listvar
- Fix toasts not doing correct HMTL here

Wolfsblvt committed:
Refactor /classify-expressions, deprecating...

- Update /classify-expressions, deprecating the old "format"
- Fix some oversights

Wolfsblvt committed:
Oversight, missed /send in refactoring...

Wolfsblvt commented:
I am too tired to update the PR description or summarize the stuff in those commits. Read the commit messages and details you can expand :/
It done though.

Cohee1207 committed:
Merge branch 'staging' into send-commands-return-value

Wolfsblvt committed:
Update doReturn() API with objectToHtmlFunc

- Allow future commands to provide a different "object to HTML" converter func than "object to text", if need be

Cohee1207 committed:
Merge branch 'send-commands-return-value' of https://github.com/SillyTavern/SillyTavern into send-commands-return-value

Cohee1207 commented on public/scripts/slash-commands/SlashCommandReturnHelper.js:
```diff
@@ -0,0 +1,80 @@
+            case 'chat-html':
+            case 'toast-text':
+            case 'toast-html': {
+                const htmlOrNotHtml = shouldHtml ? (new showdown.Converter()).makeHtml(stringValue) : escapeHtml(stringValue);
```
`/sendas return=toast-html name=Stelle <script>alert('pwned')</script>`

Wolfsblvt replied:
Done. I had it in there before. Dunno where along the way it went poof.

Cohee1207 commented:
`/listinjects` doesn't work when there are no injects in the chat. In staging it makes a toast that the chat has no injects.
<img width="546" alt="image" src="https://github.com/user-attachments/assets/26ed3b0b-27f0-49c0-8f8b-5fa1ca1beace">

Wolfsblvt committed:
sanitize HTML on html returns

- I had it in there for some time, I even tested it... likely gone during some commits

Wolfsblvt commented:
> `/listinjects` doesn't work when there are no injects in the chat. In staging it makes a toast that the chat has no injects.

You think a toast is the best way if you want pipe or object as return? But it should not crash for sure.

Edit: Fixed.

Wolfsblvt committed:
Fix empty script injections on /listinjects

Cohee1207 committed:
Merge branch 'staging' into send-commands-return-value

Cohee1207 committed:
Remove confusing toast from listvar

Cohee1207 commented:
~~`format` in `classify-expressions` is not deprecated. It defines the actual piped value. I'm fixing that.~~

UPD: !jar

Cohee1207 reviewed (approved):
Tested. I didn't try every single return type on every single command, but those that I've tried work fine.
